presto-main/src/main/java/com/facebook/presto/cost/CachingCostProvider.java

```diff
@@ -65,48 +78,221 @@ public CachingCostProvider(CostCalculator costCalculator, StatsProvider statsPro
+                .reduce(PlanCostEstimate.zero(), CachingCostProvider::addSiblingsCost);
+
+        return new PlanCostEstimate(
+                localCost.getCpuCost() + sourcesCost.getCpuCost(),
+                Math.max(sourcesCost.getMaxMemory(), Math.max(memoryEstimate.maxMemoryBeforeOutputting, memoryEstimate.maxMemoryWhenOutputting)),
+                memoryEstimate.maxMemoryWhenOutputting,
+                localCost.getNetworkCost() + sourcesCost.getNetworkCost());
+    }
+
+    private static PlanCostEstimate addSiblingsCost(PlanCostEstimate a, PlanCostEstimate b)
+    {
+        return new PlanCostEstimate(
+                a.getCpuCost() + b.getCpuCost(),
+                a.getMaxMemory() + b.getMaxMemory(),
+                a.getMaxMemoryWhenOutputting() + b.getMaxMemoryWhenOutputting(),
+                a.getNetworkCost() + b.getNetworkCost());
+    }
+
+    private static class EstimationContext
+    {
+        private final double localMaxMemory;
+        private final List<PlanCostEstimate> sourcesCosts;
+
+        public EstimationContext(double localMaxMemory, List<PlanCostEstimate> sourcesCosts)
+        {
+            this.localMaxMemory = localMaxMemory;
+            this.sourcesCosts = requireNonNull(sourcesCosts, "sourcesCosts is null");
+        }
+
+        @Override
+        public String toString()
+        {
+            return toStringHelper(this)
+                    .add("localMaxMemory", localMaxMemory)
+                    .add("sourcesCosts", sourcesCosts)
+                    .toString();
+        }
+    }
+
+    private static class MemoryEstimate
+    {
+        private final double maxMemoryBeforeOutputting;
+        private final double maxMemoryWhenOutputting;
+
+        public MemoryEstimate(double maxMemoryBeforeOutputting, double maxMemoryWhenOutputting)
+        {
+            this.maxMemoryBeforeOutputting = maxMemoryBeforeOutputting;
+            this.maxMemoryWhenOutputting = maxMemoryWhenOutputting;
+        }
+
+        @Override
+        public String toString()
+        {
+            return toStringHelper(this)
+                    .add("maxMemoryBeforeOutputting", maxMemoryBeforeOutputting)
+                    .add("maxMemoryWhenOutputting", maxMemoryWhenOutputting)
+                    .toString();
+        }
+    }
+
+    private static class PeakMemoryEstimator
+            extends PlanVisitor<MemoryEstimate, EstimationContext>
+    {
+        @Override
+        protected MemoryEstimate visitPlan(PlanNode node, EstimationContext context)
+        {
+            if (context.sourcesCosts.isEmpty()) {
+                return new MemoryEstimate(context.localMaxMemory, context.localMaxMemory);
+            }
+
+            // A PlanNode that is estimated to use memory should have an explicit rule here.
+            verify(isNaN(context.localMaxMemory) || context.localMaxMemory == 0, "Unexpected local memory usage estimation for %s: %s", node, context.localMaxMemory);
```

Comment:
what sort of plan shape would get us here?  Also, when would you get 0 vs. NaN?

Reply:
> what sort of plan shape would get us here?

I don't understand the question. 

> Also, when would you get 0 vs. NaN?

NaN is when `CostCalculatorUsingExchanges` or `CostCalculatorWithEstimatedExchanges` cannot calculate the local cost of the plan node.
0 is when they can calculate and consider it 0.

Reply:
1. I guess I'm trying to understand why all things with 0/NaN cost should have their memory output treated the same, but there shouldn't be a default for things with cost estimates.
2. If the local cost estimate is NaN, MemoryEstimate should be NaN.